### PR TITLE
[CS2113-W13-1] CurrencyManager

### PR DESCRIPTION
**Pull request by AmethystQ**

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,29 +1,134 @@
...
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|v1.0|New User|Create a wallet| can access my wallet and see its total value|
+|v1.0|User|Add money to the wallet| to increase the balance in my account|
+|v1.0|User|Withdraw money| I can indicate money has been taken from the wallet|
+|v1.0|User|get the total balance in their account|get the total amount of money in the account in their default currency|
+|v1.0|User|get the wallet details|know how much of each currency they currently have in the account|
+|v1.0|User|Request my wallet to be deleted|So that all of the data on my wallet and everything will be deleted.|
+|v1.0|User|ask for the current exchange rate between two currencies| know the rate|
+|v1.0|User|Get details about a certain currency|Find all of its symbols and the main conversion rates between the major currencies|
+|v1.0|New User|Set username and password|Sets username and password to account in order for more secure access|
```
Comment (Franky4566) [inline docs/DeveloperGuide.md:124]: Can consider Listing 'New User' after line 111 with the other 'New User' so that its easier to read

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,29 +1,134 @@
...
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### Save / Withdraw feature
+
+The save and withdraw feature is implemented based on the Wallet class data structure. 
+The Wallet class has several components: the username, the password, the default currency, the total balance, 
+and the deposit list containing different types of currency and their balances.
+
+Given below is the UML diagram of the Wallet class.
+
+![avatar](/property/WalletClass.jpg)
```
Comment (kyrenechua) [inline docs/DeveloperGuide.md:46]: Should there be an arrow to represent association from CurrencyStructure to Wallet, since Wallet has a defaultCurrency attribute.
image

Comment (malcolmang) [inline docs/DeveloperGuide.md:46]: I believe this is supposed to be a class diagram, but there are no arrows showing associations, compositions etc.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,29 +1,134 @@
...
+
+![avatar](/property/ObjectStep3.jpg)
+
+### Add/remove personal currency feature 
+
+The personal currency feature allows users to add a new currency to the list of currencies that is stored on their computer. They also have the opportunity to remove that same currency from the list so that it can’t be used anymore, but they are not able to remove any other currencies, that have been preset when downloading the jar.
+
+Given below are the UML diagrams of the classes that are needed for adding/removing a personal currency: 
+
+![avatar](/property/newcurrency1.png)
```
Comment (kyrenechua) [inline docs/DeveloperGuide.md:81]: Should there be an arrow to represent association from NewCurrency to NewCurrencyFileWorkings. Maybe instead of having many class diagrams, you can create more comprehensive class diagrams with the arrows
image

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,29 +1,134 @@
...
+![avatar](/property/WalletClass.jpg)
+
+Here is an example usage scenario and how the save/withdraw feature behaves at each step.
+
+Step 1: The user register a new account and a new `Wallet` object is initialized.
+For now, the deposit list is empty.
+
+![avatar](/property/ObjectStep1.jpg)
+
+Step 2: The user login and executes `save sgd 100` to save 100 singapore dollars in the wallet.
```
Comment (Franky4566) [inline docs/DeveloperGuide.md:55]: Consider using a sequence diagram or another diagram to explain step 2 as the parsing of inputs from method to method is quite confusing.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,29 +1,134 @@
...
+
+Step 1: The user calls the command new_currency and is then prompted in the CLI with the following commands: 
+Please enter your new currency's abbreviation: `btc`
+Please enter your new currency's full name: `bitcoin`
+Please enter your new currency's symbol: `bc`
+Please enter your new currency's rate: `0.000049`
+
+Based on that a new currency is created (in this example we have created a new currency of bitcoin).
+
+![avatar](/property/newcurrency2.png)
```
Comment (kyrenechua) [inline docs/DeveloperGuide.md:91]: Should the content after the ":" be  the types rather than the actual value? 
image

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,29 +1,134 @@
...
+
+Step 2: The user login and executes `save sgd 100` to save 100 singapore dollars in the wallet. 
+The `save` command first parses the command and verify its validity. 
+After that, it calls the `saveMoney` method of the `Wallet` class, 
+and the `saveMoney` method traverses the deposit list and finds that there's no existed SGD deposit. 
+Therefore, it creates a new `Deposit` object with the currency of SGD and the balance of 100, 
+and then adds it to the deposit list.
+Also, the total balance is updated.
+
+![avatar](/property/ObjectStep2.jpg)
```
Comment (kyrenechua) [inline docs/DeveloperGuide.md:63]: Maybe this can be the overall class diagram, with the content after the ":" being the data type rather than the example values. 
image

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,29 +1,134 @@
...
+Step 3: The user executes `withdraw sgd 20` to withdraw 20 singapore dollars in the wallet. 
+The `withdraw` command parsed the command and verify its validity. 
+Afterwards, it calls the `withdrawMoney` method of the `Wallet` class, 
+and the `withdrawMoney` method traverses the deposit list and finds that there's existing SGD deposit. 
+Therefore, it finds the corresponding `Deposit` object 
+and calls the `withdraw` method within it to withdraw 20 singapore dollars.
+Similarly, the total balance is updated.
+
+![avatar](/property/ObjectStep3.jpg)
+
```
Comment (Zhai-Yuxin) [inline docs/DeveloperGuide.md:74]: Consider using a sequence diagram for illustrating the different features

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,29 +1,134 @@
...
+Please enter your new currency's full name: `bitcoin`
+Please enter your new currency's symbol: `bc`
+Please enter your new currency's rate: `0.000049`
+
+Based on that a new currency is created (in this example we have created a new currency of bitcoin).
+
+![avatar](/property/newcurrency2.png)
+
+Step 2: In order for the new currency to be added to the currencies file, firstly the equivalent string value of it is needed to write it. For that the `getPersonalCurrencyFileLine` method of the `PersonalCurrencyList` class is called. It returns a string in the format of `“!btc,bitcoin,bc,4.9E-6”`. You can notice that there is `‘!’` in front of the string, that indicates that it is a personal currency. Then the `writeToCurrencies` method of `NewCurrencyFileWorkings` class is called with the previously returned string value as parameter to write that line to the currencies file.
+
```
Comment (Zhai-Yuxin) [inline docs/DeveloperGuide.md:94]: Consider putting diagrams for every step to show the progress like the first feature above

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (Franky4566) [inline docs/DeveloperGuide.md:46]: Diagram does not show what the 'Deposit' Class is or the connections between Wallet, Deposit and Currency Structure.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (Zhai-Yuxin) [inline docs/DeveloperGuide.md:81]: Consider showing the relationships in the class diagram instead of putting the individual classes

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,29 +1,134 @@
...
-{Describe the value proposition: what problem does it solve?}
+User who has a wallet which can have multiple currencies and can change them from one currency to another, essentially a bank account manager, where user can also withdraw and add money to the account. If they are trying to withdraw more money than they have in one currency we give them the option of exchange some from another currency (e.g trying to withdraw 100 SGD, but only have 80 SGD, we would tell them they can convert 14.34 EUR to match the remaining 20 SGD). They can give out requests to convert either one, select amount of currencies or all of the currencies they have to one specific currency. 
 
 ## User Stories
 
 |Version| As a ... | I want to ... | So that I can ...|
 |--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|v1.0|New User|Create a wallet| can access my wallet and see its total value|
```
Comment (Franky4566) [inline docs/DeveloperGuide.md:116]: Good thought to differentiate between a 'new user' and 'user'

**Review by Zhai-Yuxin — COMMENTED** [review]
Possible to add in more diagram for illustration. Good effort overall.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (Franky4566) [inline docs/DeveloperGuide.md:81]: May be clearer to add the relationships between classes in this class diagram too

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,29 +1,134 @@
...
+Please enter your new currency's abbreviation: `btc`
+Please enter your new currency's full name: `bitcoin`
+Please enter your new currency's symbol: `bc`
+Please enter your new currency's rate: `0.000049`
+
+Based on that a new currency is created (in this example we have created a new currency of bitcoin).
+
+![avatar](/property/newcurrency2.png)
+
+Step 2: In order for the new currency to be added to the currencies file, firstly the equivalent string value of it is needed to write it. For that the `getPersonalCurrencyFileLine` method of the `PersonalCurrencyList` class is called. It returns a string in the format of `“!btc,bitcoin,bc,4.9E-6”`. You can notice that there is `‘!’` in front of the string, that indicates that it is a personal currency. Then the `writeToCurrencies` method of `NewCurrencyFileWorkings` class is called with the previously returned string value as parameter to write that line to the currencies file.
```
Comment (Franky4566) [inline docs/DeveloperGuide.md:93]: Steps 2-5 are quite confusing can consider adding diagrams inbetween so that it is easier to understand

**Review by Franky4566 — COMMENTED** [review]
Overall, Good job in populating your DG! Diagram clarity can be improved.

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (zoranabc201) [inline docs/DeveloperGuide.md:46]: While some information about association is provided, it would be nice if some more information about the structure (like multiplicity, labels, etc) was provided

Comment (owenl131) [inline docs/DeveloperGuide.md:46]: Diagram should use arrows to show associations, multiplicity, directionality.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (zoranabc201) [inline docs/DeveloperGuide.md:81]: The relationships in the class diagram could be expanded upon.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,29 +1,134 @@
...
+The save and withdraw feature is implemented based on the Wallet class data structure. 
+The Wallet class has several components: the username, the password, the default currency, the total balance, 
+and the deposit list containing different types of currency and their balances.
+
+Given below is the UML diagram of the Wallet class.
+
+![avatar](/property/WalletClass.jpg)
+
+Here is an example usage scenario and how the save/withdraw feature behaves at each step.
+
```
Comment (owenl131) [inline docs/DeveloperGuide.md:49]: Scenario could be described nicely by a sequence diagram

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (zoranabc201) [inline docs/DeveloperGuide.md:91]: Shouldn't values be assigned using =? : has a special meaning in object diagrams and needs to be used carefully.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (zoranabc201) [inline docs/DeveloperGuide.md:93]: Some diagram or even code snippets would be useful here. Maybe consider avoiding a large wall of text with no illustrations as it can be a little difficult to read.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,29 +1,134 @@
...
+Step 1: The user register a new account and a new `Wallet` object is initialized.
+For now, the deposit list is empty.
+
+![avatar](/property/ObjectStep1.jpg)
+
+Step 2: The user login and executes `save sgd 100` to save 100 singapore dollars in the wallet. 
+The `save` command first parses the command and verify its validity.
```
Comment (owenl131) [inline docs/DeveloperGuide.md:56]: Could provide class diagrams for commands.

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,29 +1,134 @@
...
+  7. Click `OK` and accept the default settings.
+
 ## Acknowledgements
 
 {list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
 
 ## Design & implementation
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### Save / Withdraw feature
```
Comment (owenl131) [inline docs/DeveloperGuide.md:38]: Good job for showing how the application state changes over the course of the program.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,29 +1,134 @@
...
 ## Product scope
 ### Target user profile
 
-{Describe the target user profile}
+ Someone who works with many types of currencies, is interested in the forex market, or is interested in having a completely online wallet for tracking the money in the account.
 
 ### Value proposition
 
-{Describe the value proposition: what problem does it solve?}
+User who has a wallet which can have multiple currencies and can change them from one currency to another, essentially a bank account manager, where user can also withdraw and add money to the account. If they are trying to withdraw more money than they have in one currency we give them the option of exchange some from another currency (e.g trying to withdraw 100 SGD, but only have 80 SGD, we would tell them they can convert 14.34 EUR to match the remaining 20 SGD). They can give out requests to convert either one, select amount of currencies or all of the currencies they have to one specific currency.
```
Comment (zoranabc201) [inline docs/DeveloperGuide.md:110]: You could perhaps expand on the currency conversion feature? Right now, it allows conversion rates only once, as far as I can tell. Does it allow updating the currency? Some clarification would be nice.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (owenl131) [inline docs/DeveloperGuide.md:81]: I think it's not very clear how the different classes fit together.

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (zoranabc201) [inline docs/DeveloperGuide.md:116]: Perhaps the users in the user stories could be more specific? You could perhaps expand on who(like a bank manager, or maybe working adults, or someone else) will specifically be using your product and try to appeal to their more niche tastes.

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (owenl131) [inline docs/DeveloperGuide.md:91]: Header should be :CurrencyStructure since this is an object instance, not the class.

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,29 +1,134 @@
...
+
+### Add/remove personal currency feature 
+
+The personal currency feature allows users to add a new currency to the list of currencies that is stored on their computer. They also have the opportunity to remove that same currency from the list so that it can’t be used anymore, but they are not able to remove any other currencies, that have been preset when downloading the jar.
+
+Given below are the UML diagrams of the classes that are needed for adding/removing a personal currency: 
+
+![avatar](/property/newcurrency1.png)
+
+Step 1: The user calls the command new_currency and is then prompted in the CLI with the following commands:
```
Comment (zoranabc201) [inline docs/DeveloperGuide.md:83]: You could perhaps be clearer on what exactly are the commands that the user needs to type and what happens when the user enters that command. Maybe consider showing inputs line by line, and the response of the system, for every input rather than providing a paragraph.

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,29 +1,134 @@
...
+and the `withdrawMoney` method traverses the deposit list and finds that there's existing SGD deposit. 
+Therefore, it finds the corresponding `Deposit` object 
+and calls the `withdraw` method within it to withdraw 20 singapore dollars.
+Similarly, the total balance is updated.
+
+![avatar](/property/ObjectStep3.jpg)
+
+### Add/remove personal currency feature 
+
+The personal currency feature allows users to add a new currency to the list of currencies that is stored on their computer. They also have the opportunity to remove that same currency from the list so that it can’t be used anymore, but they are not able to remove any other currencies, that have been preset when downloading the jar.
```
Comment (zoranabc201) [inline docs/DeveloperGuide.md:77]: Maybe consider making it clearer about what exactly is the user deleting? Are they deleting a currency(as in dollars, SGD, etc) or are they deleting their own information about how much of their currency is present.

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,29 +1,134 @@
 # Developer Guide
 
+## Welcome!
+Welcome and thanks for joining us in building our program created to handle all of your currencies and money problems. With us you can use such functions as transfer, withdraw, add with different currencies to handle all your daily money problems. You also have the opportunity to add your own currencies, such as cryptocurrencies and etc. 
+
+Within this documentation you can find all the information needed to keep building it.
+
+## Setting up
```
Comment (zoranabc201) [inline docs/DeveloperGuide.md:8]: Some links to the different parts of the DG would perhaps be useful for navigation.

**Review by richwill28 — COMMENTED** [review]
[DG Review]

1. Please clean up the acknowledgements part.

image

2. While it's nice that an example scenario was provided, it's still not very clear to understand fully what's going on. Perhaps we could compliment it with a sequence diagram.

image

3. This part is perhaps too wordy, which makes it a bit difficult to read.

image

4. Just a cosmetic issue, but we can combine both tables.

image

5. Please clean up the following parts as well.

image

Overall, the document is still very lacking. Please document more things, if possible most part of the code. Remember that DG is meant for future developers should they maintain the code. So it is quite important for us to explain some of the design decisions and implementation details. Adding more (useful) diagrams would help a lot.

File: src/main/java/seedu/duke/account/MoneyCommand.java [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.List;
+
+import seedu.duke.CurrencyList;
+import seedu.duke.CurrencyStructure;
+import seedu.duke.Wallet;
+import seedu.duke.WalletFile;
+import seedu.duke.exception.FinanceException;
+import seedu.duke.exception.FinanceException.ExceptionCollection;
+
+public class MoneyCommand {
```
Comment (richwill28) [inline src/main/java/seedu/duke/account/MoneyCommand.java:10]: It's more preferred if each sub-command is a class (Single Responsibly Principle) that inherits from a more general class.

File: src/main/java/seedu/duke/Wallet.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
+package seedu.duke;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import seedu.duke.account.Deposit;
+
+public class Wallet {
+    final static CurrencyStructure INITIAL_CURRENCY = new CurrencyStructure("usd", "us dollar", "$", 1);
```
Comment (richwill28) [inline src/main/java/seedu/duke/Wallet.java]: Is this attribute use elsewhere? It's seems a bit odd as a class-level member when it's used to as an initial state in the constructor.

**Review by richwill28 — COMMENTED** [review]
[Code Review] Bao Yifan

Please implement JUnit tests.

File: src/main/java/seedu/duke/account/Account.java [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+        CurrencyStructure currency;
+        while (!isAccountExit) {
+            String in = InputManager.receiveInputLine().toLowerCase();
+            String[] splits = in.split(" ");
+
+            if (splits.length == 1) {
+                String commandType = splits[0];
+                try {
+                    switch (commandType) {
+                    case "balance":
```
Comment (richwill28) [inline src/main/java/seedu/duke/account/Account.java]: The command type could be implemented as an enumeration since it's a list of finite values. It could help more with coding, debugging and reasoning the program.

File: src/main/java/seedu/duke/account/Account.java [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+    protected Wallet wallet;
+    protected String loginTime;
+
+    public Account(Wallet wallet) {
+        this.wallet = wallet;
+        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("uuuu-MM-dd>HH:mm:ss");
+        loginTime = dtf.format(LocalDateTime.now());
+    }
+
+    public boolean handleAccountRequest() throws FinanceException {
```
Comment (richwill28) [inline src/main/java/seedu/duke/account/Account.java]: This method is too long. It's difficult to read and reason about the control flow of the program.

File: src/main/java/seedu/duke/account/Account.java [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+                }
+            }
+            else {
+                throw new FinanceException(ExceptionCollection.COMMAND_TYPE_EXCEPTION);
+            }
+        }
+        return isAccountExit;
+    }
+
+    private boolean helpCenter(){
```
Comment (richwill28) [inline src/main/java/seedu/duke/account/Account.java]: Again, this method is too long. It could be simplified into several sub-routines.

File: src/main/java/seedu/duke/AccountHistoryFile.java [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+    public static void updateLoginAccount(String username, String loginTime, String in){
+        try{
+            File f = new File(FILE_PATH + "/" + username + loginTime + ".txt");
+            FileWriter fw = new FileWriter(f, true);
+            fw.write(in);
+            fw.write(System.lineSeparator());
+            fw.close();
+        }
+        catch (IOException e){
+            //FILE NOT WORKING
```
Comment (richwill28) [inline src/main/java/seedu/duke/AccountHistoryFile.java:60]: We should probably handle the exception here?

**Review by richwill28 — COMMENTED** [review]
[Code Review] Abhitya

Please implement JUnit tests.

File: src/main/java/seedu/duke/Currency2D.java [hunk 24]
```diff
@@ -0,0 +1,99 @@
+package seedu.duke;
+
+import seedu.duke.exception.FinanceException;
+
+
+import static seedu.duke.CurrencyList.currencyList;
+import static seedu.duke.newcurrency.PersonalCurrencyList.personalCurrencyList;
+
+public class Currency2D {
+    //can add up to 30 new currencies
```
Comment (richwill28) [inline src/main/java/seedu/duke/Currency2D.java:10]: Is there any reason for the 30 currencies limit?

File: src/main/java/seedu/duke/Currency2D.java [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+    static String[] currencies = {
+            "usd,us dollar,$,1.0",
```
Comment (richwill28) [inline src/main/java/seedu/duke/Currency2D.java:12]: Instead of hard-coding each value, we could probably find a compilation somewhere. A better alternative is to replace this with an API call that is updated regularly.

File: src/main/java/seedu/duke/Currency2D.java [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+            "krw,south korean won,₩,1422.9983",
+            "lkr,sri lankan rupee,rs,366.9941",
+            "sek,swedish krona,kr,11.0958",
+            "chf,swish frank,chf,0.994",
+            "twd,taiwan new dollar,nt$,31.9973",
+            "thb,thailand bath,฿,38.0681",
+            "ttd,trinidad and tobago dollar,tt$,6.7784",
+            "try,turkish turkey,lira ₺,18.5928",
+            "vef,venezuela bolivar,bs,829225.2443",
+            "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "",
```
Comment (richwill28) [inline src/main/java/seedu/duke/Currency2D.java:66]: What are these empty string values?

File: src/main/java/seedu/duke/UserNameFileWorkings.java [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+            System.out.println(e);
+        }
+
+        try {
+            existingUserNames = getUserNames(path);
+
+        } catch (FinanceException e) {
+            createUserNames(path);
+            existingUserNames = new ArrayList<>();
+        }
```
Comment (richwill28) [inline src/main/java/seedu/duke/UserNameFileWorkings.java:43]: The two try-catches could be merged.

File: src/main/java/seedu/duke/UserNameFileWorkings.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+
+        return existingUserNames;
+    }
+
+    public static List<String> getUserNames(Path path) throws FinanceException {
```
Comment (richwill28) [inline src/main/java/seedu/duke/UserNameFileWorkings.java:48]: Any good reason why all the methods in this class are static?

**Review by richwill28 — COMMENTED** [review]
[Code Review] Markus

Please implement more JUnit tests.
